src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs

```diff
@@ -488,20 +489,108 @@ public int GetLength(MemoryPoolIterator2 end)
+            char* output = stackalloc char[length];
+
+            return SingleBlockAsciiIter(output, input, offset, length);
+        }
+
+        private static unsafe string SingleBlockAsciiIter(char* output, byte[] input, int offset, int length)
+        {
+            // avoid declaring other local vars, or doing work with stackalloc
+            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
+            for (var i = 0; i < length; i++)
+            {
+                output[i] = (char)input[i + offset];
+            }
+            return new string(output, 0, length);
+        }
+
+        private static unsafe string MultiBlockAsciiString(MemoryPoolBlock2 startBlock, ref MemoryPoolIterator2 end, int offset, int length)
+        {
+            // avoid declaring other local vars, or doing work with stackalloc
+            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
+            char* output = stackalloc char[length];
+
+            return MultiBlockAsciiIter(output, startBlock, ref end, offset, length);
+        }
+        
+        private static unsafe string MultiBlockAsciiIter(char* output, MemoryPoolBlock2 startBlock, ref MemoryPoolIterator2 end, int inputOffset, int length)
+        {
+            // avoid declaring other local vars, or doing work with stackalloc
+            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
+
+            var outputOffset = 0;
+            var block = startBlock;
+            var remaining = length;
+
+            while(true)
+            {
+                int following = (block != end._block ? block.End : end._index) - inputOffset;
+
+                if (following > 0)
+                {
+                    var input = block.Array;
+                    for (var i = 0; i < following; i++)
+                    {
+                        output[i + outputOffset] = (char)input[i + inputOffset];
+                    }
+
+                    remaining -= following;
+                    outputOffset += following;
+                }
+                
+                if (remaining == 0)
+                {
+                    return new string(output, 0, length);
+                }
+
+                block = block.Next;
+                inputOffset = block.Start;
+            }
+        }
+
+        public string GetAsciiString(ref MemoryPoolIterator2 end)
+        {
+            // avoid declaring other local vars, or doing work with stackalloc
+            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
             if (IsDefault || end.IsDefault)
             {
                 return default(string);
             }
+
+            var length = GetLength(ref end);
+
+            if (length > _maxHeaderStackLength)
+            {
+                return GetUtf8String(ref end);
```

Comment:
Seems weird to call `GetUtf8String` from `GetAsciiString`. Maybe we should call `GetUtf8String` `GetAsciiStringSlow` or something and use `Encoding.ASCII`.